### PR TITLE
feat: add a utility method for specifying static credentials

### LINKED ISSUE #274: Utility method for specifying static credentials
Sub-task of #102

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
This pr implements `auth.StaticCredential`, which is a utility method that returns a credential function that can be used in the `Credential` field of a `Client`.

See Discussion #240 
Resolves #274 

Signed-off-by: wangxiaoxuan273 <wangxiaoxuan119@gmail.com>

### CODE REVIEW

File: registry/remote/auth/example_test.go [hunk 1]
```diff
@@ -160,19 +160,11 @@ func ExampleClient_Do_minimalClient() {
...
-			default:
-				// credential not found. Try anonymous access.
-				return auth.EmptyCredential, nil
-			}
-		},
+		// expectedHostAddress is of form ipaddr:port
+		Credential: auth.StaticCredential(expectedHostAddress, auth.Credential{
+			Username: username,
+			Password: password,
+		}),
```
Comment (Wwwsylvia) [inline registry/remote/auth/example_test.go:167]: This now looks cleaner! 😄

File: registry/remote/auth/example_test.go [hunk 2]
```diff
@@ -193,18 +185,11 @@ func ExampleClient_Do_basicAuth() {
...
-		Credential: func(ctx context.Context, reg string) (auth.Credential, error) {
-			switch reg {
-			// expectedHostAddress is of form ipaddr:port
-			case expectedHostAddress:
-				return auth.Credential{
-					Username: username,
-					Password: password,
-				}, nil
-			default:
-				return auth.EmptyCredential, fmt.Errorf("credential not found for %v", reg)
```
Comment (Wwwsylvia) [inline registry/remote/auth/example_test.go:205]: Just realized that this returns an error when credential is not found for registry. Do you think we should also return error in the utility method for this case? /cc @qweeah

File: registry/remote/internal/errutil/errors.go [hunk 3]
```diff
@@ -30,6 +31,8 @@ import (
 // sufficient.
 var maxErrorBytes int64 = 8 * 1024 // 8 KiB
 
+var ErrUnauthorized error = errors.New("status unauthorized 401")
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/internal/errutil/errors.go:34]: you mean the variable name or error message?

[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/internal/errutil/errors.go:34]: error message

File: registry/remote/auth/credential.go [hunk 4]
```diff
@@ -38,3 +40,14 @@ type Credential struct {
...
 }
+
+// StaticCredential is a utility method for specifying static credentials that
+// are used when creating a client.
+func StaticCredential(registry string, cred Credential) func(ctx context.Context, reg string) (Credential, error) {
+	return func(_ context.Context, reg string) (Credential, error) {
+		if reg == registry {
+			return cred, nil
+		}
+		return EmptyCredential, nil
```
Comment (qweeah) [inline registry/remote/auth/credential.go:51]: Should add a new error like `ErrInvalidCredential` in `errdef` package and return here.

File: registry/remote/auth/credential.go [hunk 5]
```diff
@@ -38,3 +40,14 @@ type Credential struct {
 	// Reference: https://docs.docker.com/registry/spec/auth/token/
 	AccessToken string
 }
+
+// StaticCredential is a utility method for specifying static credentials that
+// are used when creating a client.
+func StaticCredential(registry string, cred Credential) func(ctx context.Context, reg string) (Credential, error) {
```
Comment (qweeah) [inline registry/remote/auth/credential.go:46]: Nit: rename `registry` to `host` and `reg` to `target`

Comment (Wwwsylvia) [reply registry/remote/auth/credential.go:46]: nice catch!

File: registry/remote/auth/credential.go (hunk 4 above)
Comment (Wwwsylvia) [reply registry/remote/auth/credential.go:51]: I think `ErrCredentialNotFound` sounds more accurate. And it is within the `auth` scope.

File: registry/remote/auth/example_test.go (hunk 2 above)
Comment (qweeah) [reply registry/remote/auth/example_test.go:205]: Yes, we should when there is an registry host name mismatch

File: registry/remote/auth/credential.go (hunk 4 above)
Comment (qweeah) [reply registry/remote/auth/credential.go:51]: Between those 2 I will still stick to `ErrInvalidCredential`. If I am a user, `ErrCredentialNotFound` sounds like that the provided static credential is somehow saved to somewhere and cannot be found.

File: registry/remote/internal/errutil/errors.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/internal/errutil/errors.go:34]: Changed.

File: registry/remote/auth/credential.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/auth/credential.go:46]: Moved.

File: registry/remote/auth/credential.go [hunk 6]
```diff
@@ -38,3 +40,14 @@ type Credential struct {
 	// Reference: https://docs.docker.com/registry/spec/auth/token/
 	AccessToken string
 }
+
+// StaticCredential is a utility method for specifying static credentials that
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/auth/credential.go:44]: Changed.

File: registry/remote/auth/credential.go (hunk 5 above)
Comment (wangxiaoxuan273) [reply registry/remote/auth/credential.go:46]: changed accordingly.

File: registry/remote/auth/credential.go (hunk 4 above)
Comment (wangxiaoxuan273) [reply registry/remote/auth/credential.go:51]: Do we have to define an error variable? Why can't we just use `fmt.Errorf("invalid credential")`?

File: registry/remote/auth/example_test.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply registry/remote/auth/example_test.go:205]: Added the error.

File: registry/remote/auth/credential.go (hunk 4 above)
Comment (wangxiaoxuan273) [reply registry/remote/auth/credential.go:51]: > Between those 2 I will still stick to `ErrInvalidCredential`. If I am a user, `ErrCredentialNotFound` sounds like that the provided static credential is somehow saved to somewhere and cannot be found.

I agree with Billy.

Comment (Wwwsylvia) [reply registry/remote/auth/credential.go]: We need an error variable and we will use that in unit testing.

File: registry/remote/auth/client.go [hunk 7]
```diff
@@ -378,6 +378,16 @@ func (c *Client) fetchOAuth2Token(ctx context.Context, realm, service string, sc
 	return "", fmt.Errorf("%s %q: empty token returned", resp.Request.Method, resp.Request.URL)
 }
 
+// StaticCredential specifies static credentials for the given registry.
+func StaticCredential(host string, cred Credential) func(ctx context.Context, target string) (Credential, error) {
+	return func(_ context.Context, target string) (Credential, error) {
+		if target == host {
+			return cred, nil
+		}
+		return EmptyCredential, fmt.Errorf("invalid credential")
```
Comment (Wwwsylvia) [inline registry/remote/auth/client.go:387]: Let's define an error variable, and we can wrap the error with an elaborated error message here.

Comment (wangxiaoxuan273) [reply registry/remote/auth/client.go]: Done.

File: registry/remote/auth/client.go [hunk 8]
```diff
@@ -378,6 +378,18 @@ func (c *Client) fetchOAuth2Token(ctx context.Context, realm, service string, sc
...
 
+var ErrInvalidCredential error = errors.New("invalid credential")
+
+// StaticCredential specifies static credentials for the given registry.
+func StaticCredential(host string, cred Credential) func(ctx context.Context, target string) (Credential, error) {
+	return func(_ context.Context, target string) (Credential, error) {
+		if target == host {
+			return cred, nil
+		}
+		return EmptyCredential, ErrInvalidCredential
```
Comment (Wwwsylvia) [inline registry/remote/auth/client.go:389]: nit: Do we want to include the host name in the error message?

Comment (wangxiaoxuan273) [reply registry/remote/auth/client.go:389]: Added the target name to the error message.

**Review by Wwwsylvia — APPROVED** [review]
LGTM

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: registry/remote/internal/errutil/errors.go [hunk 9]
```diff
@@ -80,5 +83,9 @@ func ParseErrorResponse(resp *http.Response) error {
 	} else {
 		errmsg = http.StatusText(resp.StatusCode)
 	}
+
+	if resp.StatusCode == http.StatusUnauthorized {
+		return fmt.Errorf("%s %q: unexpected status code %d: %s, error = %w", resp.Request.Method, resp.Request.URL, resp.StatusCode, errmsg, ErrUnauthorized)
+	}
 	return fmt.Errorf("%s %q: unexpected status code %d: %s", resp.Request.Method, resp.Request.URL, resp.StatusCode, errmsg)
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply registry/remote/internal/errutil/errors.go:90]: Later you can try `error.As`.

Here's an example:
```go
func main() {
	err := &UnexpectedStatusCodeError{
		Method:     "GET",
		URL:        &url.URL{},
		StatusCode: 401,
		Message:    "unauthorized",
	}
	wrappedError := fmt.Errorf("wrap error: %w", err)

	var unwrappedError *UnexpectedStatusCodeError
	if errors.As(wrappedError, &unwrappedError) {
		fmt.Println("Failed status:", unwrappedError.StatusCode)
	} else {
		fmt.Println(wrappedError)
	}
}
```

File: registry/remote/auth/client.go [hunk 10]
```diff
@@ -52,6 +52,20 @@ var maxResponseBytes int64 = 128 * 1024 // 128 KiB
...
+// when no valid credential is found given the target host.
+var ErrInvalidCredential error = errors.New("invalid credential")
+
+// StaticCredential specifies static credentials for the given host.
+func StaticCredential(host string, cred Credential) func(ctx context.Context, target string) (Credential, error) {
+	return func(_ context.Context, target string) (Credential, error) {
+		if target == host {
+			return cred, nil
+		}
+		return EmptyCredential, fmt.Errorf("no valid credential found for target %s: %w", target, ErrInvalidCredential)
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/auth/client.go:65]: But then the returned error is always `nil`. Do we even need to return the error?

[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/auth/client.go:65]: Changed accordingly.

File: registry/remote/auth/client.go [hunk 11]
```diff
@@ -52,6 +52,20 @@ var maxResponseBytes int64 = 128 * 1024 // 128 KiB
 // See also ClientID.
 var defaultClientID = "oras-go"
 
+// ErrInvalidCredential will be returned by a client's Credential method
+// when no valid credential is found given the target host.
+var ErrInvalidCredential error = errors.New("invalid credential")
+
+// StaticCredential specifies static credentials for the given host.
+func StaticCredential(host string, cred Credential) func(ctx context.Context, target string) (Credential, error) {
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/auth/client.go:60]: Changed accordingly.

File: registry/remote/internal/errutil/errors.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/internal/errutil/errors.go]: Changed accordingly.

**Pull request merged**